src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImporter.cs

```diff
@@ -85,135 +85,21 @@ await foreach (ImportResource resource in inputChannel.Reader.ReadAllAsync(cance
 
         private void ImportResourcesInBuffer(List<ImportResource> resources, List<string> errors, ImportMode importMode, CancellationToken cancellationToken, ref long succeededCount, ref long failedCount, ref long processedBytes)
         {
-            var retries = 0;
-            var loaded = new List<ImportResource>();
-            var conflicts = new List<ImportResource>();
-            while (true)
-            {
-                try
-                {
-                    loaded = new List<ImportResource>();
-                    conflicts = new List<ImportResource>();
-                    ImportResourcesInBufferInternal(resources, loaded, conflicts, importMode, retries == 0, cancellationToken).Wait();
-                    break;
-                }
-                catch (Exception e)
-                {
-                    var sqlEx = (e is SqlException ? e : e.InnerException) as SqlException;
-                    if (sqlEx != null && sqlEx.Number == SqlErrorCodes.Conflict && retries++ < 30)
-                    {
-                        _logger.LogWarning(e, $"Error on {nameof(ImportResourcesInBufferInternal)} retries={{Retries}}", retries);
-                        Task.Delay(5000, cancellationToken);
-                        continue;
-                    }
-
-                    _logger.LogError(e, $"Error on {nameof(ImportResourcesInBufferInternal)} retries={{Retries}}", retries);
-                    _store.StoreClient.TryLogEvent(nameof(ImportResourcesInBufferInternal), "Error", $"retries={retries} error={e}", null, cancellationToken).Wait();
-
-                    throw;
-                }
-            }
-
-            var errorResources = resources.Where(r => !string.IsNullOrEmpty(r.ImportError));
-            errors.AddRange(errorResources.Select(r => r.ImportError));
-            var dups = resources.Where(r => string.IsNullOrEmpty(r.ImportError)).Except(loaded).Except(conflicts);
-            AppendErrorsToBuffer(dups, conflicts, errors);
-
-            succeededCount += loaded.Count;
-            failedCount += errorResources.Count() + dups.Count() + conflicts.Count;
+            errors.AddRange(resources.Where(r => !string.IsNullOrEmpty(r.ImportError)).Select(r => r.ImportError));
+            failedCount += errors.Count;
+            //// exclude resources with parsing error (ImportError != null)
+            var validResources = resources.Where(r => string.IsNullOrEmpty(r.ImportError)).ToList();
+            var results = _store.ImportResourcesAsync(validResources, importMode, cancellationToken).Result;
```

Comment:
Calling "Result" here will most likely run into a deadlock at some point. It isn't recommended to do this. I realize in the previous version of the code it was calling "Wait" and not returning any values.  I would recommend making the calling method async and follow that up through the code to ensure it doesn't run into issues. If you call "Result" here without any kind of wait there is a high chance it won't have the method call complete, so your results will be inconsistent.

Syntactically the old code and this refactor look similar functionally, but they will behave differently when ran.

Reply:
OK. Fixed.